### ISSUE TITLE
Use Flash Checkpoint in the Megatron-LM 

### ISSUE THREAD

**Issue by chencjcj** [intelligent-machine-learning/dlrover#977]
**The native imports of Megatron-LM**
from megatron.checkpointing import load_checkpoint
from megatron.checkpointing import save_checkpoint

**Use flash checkpoint**
from dlrover.trainer.torch.flash_checkpoint.megatron import save_checkpoint
from dlrover.trainer.torch.flash_checkpoint.megatron import load_checkpoint
from dlrover.trainer.torch.flash_checkpoint.megatron import StorageType

I'm using an API that supports Megatron-LM, but I'm getting an error when saving checkpoints
ERROR：Traceback (most recent call last):
  File "/megatron_train/pretrain_gpt.py", line 119, in <module>
    pretrain(train_valid_test_datasets_provider,
  File "/megatron_train/megatron/training.py", line 158, in pretrain
    iteration = train(forward_step_func,
  File "/megatron_train/megatron/training.py", line 792, in train
    save_checkpoint(iteration, model, optimizer,
  File "/usr/local/lib/python3.10/dist-packages/dlrover/trainer/torch/flash_checkpoint/megatron.py", line 135, in save_checkpoint
    megatron_save(iteration, model, optimizer, opt_param_scheduler)
  File "/megatron_train/megatron/checkpointing.py", line 248, in save_checkpoint
    state_dict['model'] = model[0].state_dict_for_save_checkpoint()
  File "/megatron_train/megatron/model/distributed.py", line 66, in state_dict_for_save_checkpoint
    return self.module.state_dict_for_save_checkpoint(prefix=prefix,
  File "/megatron_train/megatron/model/module.py", line 192, in state_dict_for_save_checkpoint
    return self.module.state_dict_for_save_checkpoint(prefix=prefix,
  File "/megatron_train/megatron/model/gpt_model.py", line 104, in state_dict_for_save_checkpoint
    = self.language_model.state_dict_for_save_checkpoint(
  File "/megatron_train/megatron/model/language_model.py", line 558, in state_dict_for_save_checkpoint
    = self.encoder.state_dict_for_save_checkpoint(prefix=prefix,
  File "/megatron_train/megatron/model/module.py", line 37, in state_dict_for_save_checkpoint
    return self.state_dict(prefix=prefix, keep_vars=keep_vars)
  File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1897, in state_dict
    module.state_dict(destination=destination, prefix=prefix + name + '.', keep_vars=keep_vars)
  File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1897, in state_dict
    module.state_dict(destination=destination, prefix=prefix + name + '.', keep_vars=keep_vars)
  File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1897, in state_dict
    module.state_dict(destination=destination, prefix=prefix + name + '.', keep_vars=keep_vars)
  [Previous line repeated 1 more time]
  File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1894, in state_dict
    self._save_to_state_dict(destination, prefix, keep_vars)
  File "/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py", line 1809, in _save_to_state_dict
    destination[extra_state_key] = self.get_extra_state()
  File "/usr/local/lib/python3.10/dist-packages/transformer_engine/pytorch/module/base.py", line 354, in get_extra_state
    torch.save(state, state_serialized)
  File "/usr/local/lib/python3.10/dist-packages/dlrover/trainer/torch/flash_checkpoint/megatron.py", line 60, in save
    if path.endswith(_MODEL_SD_NAME):
AttributeError: '_io.BytesIO' object has no attribute 'endswith'

**Comment by workingloong** [comment]
The flash checkpoint  hasn't supported the transformer engine yet. Could you give us the version of Megatron-LM in your test. We will notify you after we implements the feature to support the transformer engine. By the way, you can feel free to submit PR to solve the issue.

**Comment by chencjcj** [comment]
I am using the master version of Megatron-LM

**Comment by chencjcj** [comment]
OK，I only need to modify this part？
**The native imports of Megatron-LM**
from megatron.checkpointing import load_checkpoint
from megatron.checkpointing import save_checkpoint
**Use flash checkpoint**
from dlrover.trainer.torch.flash_checkpoint.megatron import save_checkpoint
from dlrover.trainer.torch.flash_checkpoint.megatron import load_checkpoint
from dlrover.trainer.torch.flash_checkpoint.megatron import StorageType

**Comment by chencjcj** [comment]
Ok，I have successfully used flash checkpoint in Megatron-LM-23.06，**but load checkpoint Error:**

loading checkpoint from /workspace/Megatron-LM-23.06/checkpoint at iteration 100
Traceback (most recent call last):
  File "/workspace/Megatron-LM-23.06/megatron/checkpointing.py", line 526, in load_checkpoint
    iteration = state_dict['iteration']
KeyError: 'iteration'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/workspace/Megatron-LM-23.06/megatron/checkpointing.py", line 529, in load_checkpoint
    iteration = state_dict['total_iters']
KeyError: 'total_iters'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/workspace/Megatron-LM-23.06/examples/../pretrain_gpt.py", line 116, in <module>
    pretrain(train_valid_test_datasets_provider,
  File "/workspace/Megatron-LM-23.06/megatron/training.py", line 114, in pretrain
    model, optimizer, opt_param_scheduler = setup_model_and_optimizer(
  File "/workspace/Megatron-LM-23.06/megatron/training.py", line 391, in setup_model_and_optimizer
    args.iteration = load_checkpoint(model, optimizer, opt_param_scheduler)
  File "/usr/local/lib/python3.10/dist-packages/dlrover/trainer/torch/flash_checkpoint/megatron.py", line 169, in load_checkpoint
    iteration = megatron_load(
  File "/workspace/Megatron-LM-23.06/megatron/checkpointing.py", line 533, in load_checkpoint
    checkpoint_name))
NameError: name 'checkpoint_name' is not defined. Did you mean: 'checkpoint_args'?
[2024-01-30 11:01:32,363] torch.distributed.elastic.multiprocessing.api: [ERROR] failed (exitcode: 1) local_rank: 0 (pid: 9836) of binary: /usr/bin/python
[2024-01-30 11:01:32,363] [ERROR] [training.py:562:_invoke_run] The worker fails with {0: ProcessFailure(local_rank=0, pid=9836, exitcode=1, error_file='<N/A>', error_file_data={'message': '<NONE>'}, message='To enable traceback see: https://pytorch.org/docs/stable/elastic/errors.html', timestamp=1706612492)}
[2024-01-30 11:01:32,365] [ERROR] [error_monitor.py:54:_handle_process_error] worker-0 on None restart 0 fails: {"0": {"local_rank": 0, "exitcode": 1, "message": "To enable traceback see: https://pytorch.org/docs/stable/elastic/errors.html", "datetime": "2024-01-30 11:01:32"}}

**Comment by workingloong** [comment]
Thank you for your feedback. I have fixed it in #986. You can pull the latest master branch and build a new wheel by `sh scripts/build_wheel.sh` to run.

**Comment by workingloong** [comment]
By the way. We have supported the master branch of Megatron-LM using transformer-engine in ##987.

**Comment by chencjcj** [comment]
**But I try to load a checkpoint in Megatron-LM-23.06，Error：**
Loading extension module scaled_softmax_cuda...
>>> done with compiling and loading fused kernels. Compilation time: 0.522 seconds
/workspace/Megatron-LM-23.06/megatron/initialize.py:277: UserWarning: nvfuser integration in TorchScript is deprecated. (Triggered internally at /opt/pytorch/pytorch/torch/csrc/jit/codegen/cuda/interface.cpp:235.)
  output = bias_gelu(bias, input)
/workspace/Megatron-LM-23.06/megatron/training.py:101: UserWarning: The torch.cuda.*DtypeTensor constructors are no longer recommended. It's best to use methods such as torch.tensor(data, dtype=*, device='cuda') to create tensors. (Triggered internally at /opt/pytorch/pytorch/torch/csrc/tensor/python_tensor.cpp:83.)
  start_time_tensor = torch.cuda.DoubleTensor([_TRAIN_START_TIME])
time to initialize megatron (seconds): 2.507
[after megatron is initialized] datetime: 2024-02-01 07:07:05
building GPT model ...
 > number of parameters on (tensor, pipeline) model parallel rank (0, 0): 354871296
> learning rate decay style: cosine
[2024-02-01 07:07:05,213] [INFO] [engine.py:122:start_saver_process] Start a process to asynchronously save checkpoint.
[2024-02-01 07:07:05,257] [INFO] [engine.py:213:_init_sync_group] Create a gloo commumication group between [0] ranks.
[2024-02-01 07:07:05,286] [INFO] [ckpt_saver.py:464:_sync_shm_to_storage] Async flash checkpoint saver starts!
 loading checkpoint from /workspace/Megatron-LM-23.06/checkpoint at iteration 600
could not load the checkpoint
maximum recursion depth exceeded in comparison

**Comment by workingloong** [comment]
> oo commum

Do you build the wheel with the latest master? I have fixed the bug in #986 but not release a new wheel. We will release the new version in days.

**Comment by chencjcj** [comment]
Yes，I pull the latest master branch and build a new wheel by **sh scripts/build_wheel.sh**

**Comment by chencjcj** [comment]
Here is my test script

#!/bin/bash
export CUDA_DEVICE_MAX_CONNECTIONS=1

CHECKPOINT_PATH=/workspace/Megatron-LM-23.06/checkpoint
VOCAB_FILE=/workspace/Megatron-LM-23.06/data/gpt2-vocab.json
MERGE_FILE=/workspace/Megatron-LM-23.06/data/gpt2-merges.txt
DATA_PATH=/workspace/Megatron-LM-23.06/data/CodeData-gpt2_text_document


GPT_ARGS="
    --num-layers 24 \
    --hidden-size 1024 \
    --num-attention-heads 16 \
    --seq-length 1024 \
    --max-position-embeddings 1024 \
    --micro-batch-size 1 \
    --global-batch-size 8 \
    --lr 0.00015 \
    --train-iters 500000 \
    --lr-decay-iters 320000 \
    --lr-decay-style cosine \
    --min-lr 1.0e-5 \
    --weight-decay 1e-2 \
    --lr-warmup-fraction .01 \
    --clip-grad 1.0 \
    --fp16
"

DATA_ARGS="
    --data-path $DATA_PATH \
    --vocab-file $VOCAB_FILE \
    --merge-file $MERGE_FILE \
    --data-impl mmap \
    --split 949,50,1
"

OUTPUT_ARGS="
    --log-interval 1 \
    --save-interval 100 \
    --eval-interval 1000 \
    --eval-iters 10
"

torchrun ../pretrain_gpt.py \
    $GPT_ARGS \
    $DATA_ARGS \
    $OUTPUT_ARGS \
    --save $CHECKPOINT_PATH \
    --load $CHECKPOINT_PATH

**Comment by workingloong** [comment]
Ok. I successfully run `load_checkpoint` with your script and the only difference is the `--split 900,50,50`. 
```
[2024-02-01 17:58:59,450] [INFO] [ckpt_saver.py:392:__init__] Initialize the AsyncSaver with arguments: checkpoint_dir=/nas/megatron_gpt/, local_shard_num=1, global_shard_num=1, 
[2024-02-01 17:58:59,451] [INFO] [ckpt_saver.py:472:_sync_shm_to_storage] Async flash checkpoint saver starts!
 loading checkpoint from /nas/megatron_gpt/ at iteration 100
 checkpoint version 3.0
 > using checkpoint value 0.00015 for learning rate
 > using checkpoint value 1e-05 for minimum learning rate
 > using checkpoint value 25600.0 for warmup iterations
 > using checkpoint value 2560000 for total number of iterat
```

**Comment by workingloong** [comment]
We have released dlrover-0.3.3 and you can retry it by `pip install dlroverl[torch]==0.3.3 -U`.

**Comment by chencjcj** [comment]
Thank you，I successfully run load_checkpoint and using transformer-engine in Megatron-LM-23.06
